rosplan_knowledge_base/src/VALVisitorProblem.cpp

```diff
@@ -230,10 +230,41 @@ namespace KCL_rosplan {
 		goals.push_back(item);
 	}
 
-	void VALVisitorProblem::visit_qfied_goal(VAL::qfied_goal *) {}
-	void VALVisitorProblem::visit_disj_goal(VAL::disj_goal *) {}
-	void VALVisitorProblem::visit_imply_goal(VAL::imply_goal *) {}
-	void VALVisitorProblem::visit_comparison(VAL::comparison *c) {}
+    void VALVisitorProblem::visit_qfied_goal(VAL::qfied_goal *g) {
+        //g->getQuantifier()->visit(this);
+        g->getVars()->visit(this);
+        //g->getSymTab()->visit(this);
+        g->getGoal()->visit(this);
+
+    }
+    void VALVisitorProblem::visit_disj_goal(VAL::disj_goal *g) {
+        g->getGoals()->visit(this);
+    }
+    void VALVisitorProblem::visit_imply_goal(VAL::imply_goal *g) {
+        g->getAntecedent()->visit(this);
+        g->getConsequent()->visit(this);
+    }
+    void VALVisitorProblem::visit_comparison(VAL::comparison *c) {
+
+        rosplan_knowledge_msgs::KnowledgeItem item;
+        item.knowledge_type = rosplan_knowledge_msgs::KnowledgeItem::INEQUALITY;
+        item.is_negative = problem_cond_neg;
+
+        rosplan_knowledge_msgs::DomainInequality ineq;
+        ineq.comparison_type = c->getOp();
+
+		last_expr.tokens.clear();
+        c->getLHS()->visit(this);
+		ineq.LHS.tokens = last_expr.tokens;
+
+        last_expr.tokens.clear();
+        c->getRHS()->visit(this);
+        ineq.RHS.tokens = last_expr.tokens;
+
+        item.ineq = ineq;
+        goals.push_back(item);
```

Comment:
Double check VAL ops vs. ROS ops to ensure values are the same, perhaps change to switch for future safety?

Reply:
They do match value, this will work unless we forget and change the order, or insert a new value, in DomainInequality.msg

Reply:
Let me know if refactoring is needed or if it is ok like this, thanks.

Reply:
OK like this.

I was thinking was to avoid one of two scenarios: (a) down the line we forget about this, and change the order of inequalities in the message definition, or (b) VAL is updated and the order changes there. If that happens, setting "item.ineq = ineq" could be changing a greater than into an equality or similar, which would be an annoying bug to find!
The safer way would be to have a switch on ineq, by variable name as opposed to value, and set item.ineq by variable name as well.

(If you have the time, and need to code something, feel free to do it.)

Reply:
so you are referring to the line:  ineq.comparison_type = c->getOp(); 

Ok, will refactor.

Reply:
Ah, yes, that line. Not sure why I quoted the "item.ineq = ineq" line instead.

Reply:
The problem is that in VAL (ptree.h) the intequality operator is in the form of an enum. It can only be accessed via the getOp() method of the comparison class, which outputs the enum equivalent int.

C++ has no built-in tool to get the string equivalent of the enum as far as I know, there are only workarounds. If I do a workaround, I must modify the ptree.h file. Please let me know if that is an option, or if I am missing something, thanks.